### PR TITLE
[reward] fix: mask tool observations before default reward scoring

### PR DESCRIPTION
**Pull request by haoyang9804**
### What does this PR do?

This PR fixes a quiet reward false-positive bug in multi-turn/tool rollout.

Default reward managers such as `naive`, `dapo`, and `batch` decode the response
with `attention_mask`. In agent/tool rollout, `attention_mask` marks both
model-generated tokens and tool observation tokens as present, while
`response_mask` marks only policy-generated tokens. A reward function can
therefore see tool output as if it were assistant output.

Concrete trigger:

```text
prompts:        [prompt]
responses:      [assistant, final, TOOL_LEAK, pad]
attention_mask: [prompt=1, assistant=1, final=1, TOOL_LEAK=1, pad=0]
response_mask:  [assistant=1, final=1, TOOL_LEAK=0, pad=0]
```

Before this fix, default reward scoring can pass this `solution_str` to the
reward function:

```text
assistant final TOOL_LEAK
```

That is wrong for policy reward: `TOOL_LEAK` is a tool observation and is not
optimized by the policy loss. For example, if a verifier/tool observation
contains the final answer, the model can receive positive reward for text it did
not generate.

After this fix, default reward managers pass only model-generated response
tokens to `solution_str`:

```text
assistant final
```

They also place the scalar reward on the last model-generated token instead of
the last attention-valid tool token. The full decoded trajectory remains
available as `extra_info["full_response_str"]` for custom reward functions that
intentionally need tool context.

### Root Cause

Default reward managers used `attention_mask[:, prompt_len:]` to decide which
response tokens to decode and where to place scalar reward.

In single-turn rollout this is fine because attention-valid response tokens are
model tokens. In multi-turn/tool rollout, tool observations are also
attention-valid. The correct mask for policy reward text and reward placement is
`response_mask` when it is present.

### Fix

This PR adds `select_response_ids_for_reward()` and uses it in the default
reward managers.

- If `response_mask` is absent, keep the existing attention-mask behavior.
- If `response_mask` is present, decode `solution_str` from model-generated
  response tokens only.
- Place scalar reward on the last model-generated token.
- Preserve the full attention-valid decoded response in
  `extra_info["full_response_str"]`.

### Duplicate Check

I checked upstream PRs and issues before preparing this patch, following
`AGENTS.md`.

- PR query: `reward manager response_mask tool observations solution_str`
  - Link: https://github.com/verl-project/verl/pulls?q=is%3Apr+reward+manager+response_mask+tool+observations+solution_str
  - Result: no matching upstream PR.
- PR query: `tool observation reward manager response_mask`
  - Link: https://github.com/verl-project/verl/pulls?q=is%3Apr+tool+observation+reward+manager+response_mask
  - Result: returned unrelated feature work such as #4897, but no PR that fixes
    default reward managers decoding `solution_str` from tool observations.
- PR query: `reward manager tool response response_mask`
  - Link: https://github.com/verl-project/verl/pulls?q=is%3Apr+reward+manager+tool+response+response_mask
  - Result: returned broader multi-turn/agent-loop PRs such as #5506, #5951,
    #5503, #4897, #4096, #4176, #2113, and #1037. These are not duplicates:
    none patches `naive`, `dapo`, or `batch` reward managers to use
    `response_mask` for `solution_str` and reward placement.
- Issue query: `tool observations reward manager response_mask`
  - Link: https://github.com/verl-project/verl/issues?q=is%3Aissue+tool+observations+reward+manager+response_mask
  - Result: no exact open issue for default reward managers scoring tool
    observations as assistant output.

Related but not duplicate:

- Issue #4814 asks how custom reward functions can access the full rollout. This
  PR does not remove access to full rollout text; it keeps that text in
  `extra_info["full_response_str"]`. The fix only changes the default
  policy-reward `solution_str` so it no longer treats `response_mask=0` tool
  observations as model output.

### Validation Recipe

```json
{
  "kind": "rl_sentinel_validation_recipe",
  "schema_version": 1,
  "bug_id": "BUG-42F0E8A5C9D1",
  "title": "Default reward managers score tool observations as model output",
  "target": "verl",
  "validation_mode": "real_target_boundary_hook",
  "target_boundaries": [
    "verl.workers.reward_manager.naive.NaiveRewardManager.__call__"
  ],
  "requires_model_download": false,
  "requires_dataset_download": false,
  "environment": {
    "VERL_REPO": "path to the veRL checkout under test",
    "OUTPUT_DIR": "directory where validation JSON should be written",
    "EXPECTATION": "confirmed for unpatched, not_triggered for fixed",
    "PYTHON": "python3"
  },
  "constructed_scenario": {
    "reward_manager": "naive",
    "prompt_ids": [1],
    "response_ids": [10, 11, 99, 0],
    "decoded_response_ids": {
      "10": "assistant",
      "11": "final",
      "99": "TOOL_LEAK",
      "0": "<pad>"
    },
    "attention_mask": [1, 1, 1, 1, 0],
    "response_mask": [1, 1, 0, 0],
    "reward_function": "returns -1.0 if TOOL_LEAK is present in solution_str, else 1.0"
  },
  "expected_unpatched": {
    "status": "confirmed",
    "solution_str": "assistant final TOOL_LEAK",
    "reward_tensor": [[0.0, 0.0, -1.0, 0.0]]
  },
  "expected_fixed": {
    "status": "not_triggered",
    "solution_str": "assistant final",
    "full_response_str": "assistant final TOOL_LEAK",
    "reward_tensor": [[0.0, 1.0, 0.0, 0.0]]
  },
  "outputs": {
    "validation": "training_signal_validation.json"
  }
}
```

### Validation Runner

```bash
#!/usr/bin/env bash
set -euo pipefail

: "${VERL_REPO:?Set VERL_REPO to the veRL checkout under test}"
SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
OUTPUT_DIR="${OUTPUT_DIR:-${SCRIPT_DIR}/output}"
PYTHON="${PYTHON:-python3}"
EXPECTATION="${EXPECTATION:-any}"

mkdir -p "${OUTPUT_DIR}"

export VERL_REPO OUTPUT_DIR EXPECTATION
export PYTHONPATH="${VERL_REPO}${PYTHONPATH:+:${PYTHONPATH}}"

"${PYTHON}" "${SCRIPT_DIR}/reward_manager_tool_observation_hook.py"
```

Run against unpatched and fixed checkouts:

```bash
VERL_REPO=/path/to/unpatched-verl \
OUTPUT_DIR=./BUG-42F0E8A5C9D1-unpatched \
EXPECTATION=confirmed \
./run_validation.sh

VERL_REPO=/path/to/fixed-verl \
OUTPUT_DIR=./BUG-42F0E8A5C9D1-fixed \
EXPECTATION=not_triggered \
./run_validation.sh
```

### Boundary Hook

```python
from __future__ import annotations

import json
import os
import subprocess
import sys
from pathlib import Path

import numpy as np
import torch


BUG_ID = "BUG-42F0E8A5C9D1"
TITLE = "Default reward managers score tool observations as model output"


class ToyTokenizer:
    def __init__(self) -> None:
        self.vocab = {
            1: "prompt",
            10: "assistant",
            11: "final",
            99: "TOOL_LEAK",
            0: "<pad>",
        }

    def decode(self, token_ids, skip_special_tokens=True):
        ids = token_ids.tolist() if hasattr(token_ids, "tolist") else list(token_ids)
        pieces = [self.vocab[int(token_id)] for token_id in ids]
        if skip_special_tokens:
            pieces = [piece for piece in pieces if piece != "<pad>"]
        return " ".join(pieces)


def _git_commit(repo: Path) -> str | None:
    try:
        return subprocess.check_output(["git", "-C", str(repo), "rev-parse", "HEAD"], text=True).strip()
    except Exception:
        return None


def _tolist(tensor: torch.Tensor) -> list:
    return tensor.detach().cpu().tolist()


def _write_json(path: Path, payload: dict) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(payload, indent=2) + "\n", encoding="utf-8")


def main() -> None:
    repo = Path(os.environ["VERL_REPO"]).resolve()
    output_dir = Path(os.environ["OUTPUT_DIR"]).resolve()
    expectation = os.environ.get("EXPECTATION", "any")

    if str(repo) not in sys.path:
        sys.path.insert(0, str(repo))

    from verl import DataProto
    from verl.workers.reward_manager.naive import NaiveRewardManager

    captured = {}

    def compute_score(data_source, solution_str, ground_truth, extra_info):
        captured["data_source"] = data_source
        captured["solution_str"] = solution_str
        captured["ground_truth"] = ground_truth
        captured["extra_info"] = dict(extra_info)
        return -1.0 if "TOOL_LEAK" in solution_str else 1.0

    data = DataProto.from_dict(
        tensors={
            "prompts": torch.tensor([[1]], dtype=torch.long),
            "responses": torch.tensor([[10, 11, 99, 0]], dtype=torch.long),
            "attention_mask": torch.tensor([[1, 1, 1, 1, 0]], dtype=torch.long),
            "response_mask": torch.tensor([[1, 1, 0, 0]], dtype=torch.long),
        },
        non_tensors={
            "data_source": np.array(["tool_task"], dtype=object),
            "reward_model": np.array([{"ground_truth": "final"}], dtype=object),
            "extra_info": np.array([{}], dtype=object),
        },
    )

    manager = NaiveRewardManager(tokenizer=ToyTokenizer(), num_examine=0, compute_score=compute_score)
    reward_tensor = manager(data)

    solution_str = captured.get("solution_str")
    full_response_str = captured.get("extra_info", {}).get("full_response_str")
    reward = _tolist(reward_tensor)

    if solution_str == "assistant final TOOL_LEAK" and reward == [[0.0, 0.0, -1.0, 0.0]]:
        status = "confirmed"
        finding = "default reward manager scored a response_mask=0 tool observation as model output"
    elif solution_str == "assistant final" and full_response_str == "assistant final TOOL_LEAK" and reward == [
        [0.0, 1.0, 0.0, 0.0]
    ]:
        status = "not_triggered"
        finding = "default reward manager used response_mask for reward text and reward placement"
    else:
        status = "unexpected"
        finding = "reward manager produced an unrecognized response/reward shape"

    payload = {
        "kind": "rl_sentinel_training_signal_validation",
        "bug_id": BUG_ID,
        "title": TITLE,
        "status": status,
        "finding": finding,
        "commit": _git_commit(repo),
        "target_boundaries": [
            "verl.workers.reward_manager.naive.NaiveRewardManager.__call__",
        ],
        "scenario": {
            "prompt_ids": [[1]],
            "response_ids": [[10, 11, 99, 0]],
            "attention_mask": [[1, 1, 1, 1, 0]],
            "response_mask": [[1, 1, 0, 0]],
            "token_text": {
                "1": "prompt",
                "10": "assistant",
                "11": "final",
                "99": "TOOL_LEAK",
                "0": "<pad>",
            },
            "reward_rule": "return -1.0 if TOOL_LEAK appears in solution_str, else 1.0",
        },
        "observed": {
            "solution_str": solution_str,
            "full_response_str": full_response_str,
            "reward_tensor": reward,
            "score_data_source": captured.get("data_source"),
            "score_ground_truth": captured.get("ground_truth"),
        },
    }

    output_path = output_dir / "training_signal_validation.json"
    _write_json(output_path, payload)
    print(json.dumps(payload, indent=2))

    if expectation != "any" and status != expectation:
        raise SystemExit(f"expected status {expectation!r}, got {status!r}")


if __name__ == "__main__":
    main()
```

### Validation Output

Unpatched checkout:

```json
{
  "bug_id": "BUG-42F0E8A5C9D1",
  "status": "confirmed",
  "finding": "default reward manager scored a response_mask=0 tool observation as model output",
  "observed": {
    "solution_str": "assistant final TOOL_LEAK",
    "full_response_str": null,
    "reward_tensor": [[0.0, 0.0, -1.0, 0.0]]
  }
}
```

Fixed branch:

```json
{
  "bug_id": "BUG-42F0E8A5C9D1",
  "status": "not_triggered",
  "finding": "default reward manager used response_mask for reward text and reward placement",
  "observed": {
    "solution_str": "assistant final",
    "full_response_str": "assistant final TOOL_LEAK",
    "reward_tensor": [[0.0, 1.0, 0.0, 0.0]]
  }
}
```

### Test

```bash
python3 -m pytest -q tests/workers/reward_manager/test_response_mask_on_cpu.py tests/workers/reward_manager/test_registry_on_cpu.py
python3 -m ruff check verl/workers/reward_manager/response_utils.py verl/workers/reward_manager/naive.py verl/workers/reward_manager/dapo.py verl/workers/reward_manager/batch.py tests/workers/reward_manager/test_response_mask_on_cpu.py
python3 -m ruff format --check verl/workers/reward_manager/response_utils.py verl/workers/reward_manager/naive.py verl/workers/reward_manager/dapo.py verl/workers/reward_manager/batch.py tests/workers/reward_manager/test_response_mask_on_cpu.py
```

Results:

- `8 passed, 2 warnings in 7.64s`
- `All checks passed!`
- `5 files already formatted`

### API and Usage Example

No public API or config change.

Existing default reward managers now treat `response_mask` as the policy-reward
visibility mask when it is present. Custom reward functions that need the full
tool trajectory can read `extra_info["full_response_str"]`.

### Checklist Before Starting

- [x] Read `CONTRIBUTING.md`.
- [x] Read `AGENTS.md`.
- [x] Searched upstream PRs/issues and documented duplicate-check results above.
- [x] Format the PR title as `[{modules}] {type}: {description}`.

### Checklist Before Submitting

- [x] Added focused CPU regression coverage for masked tool observations.
- [x] Ran focused pytest and ruff checks listed above.
- [ ] Run full `pre-commit run --all-files --show-diff-on-failure --color=always ruff` before requesting review.
- [ ] Request CI after opening the PR.
- [ ] Human submitter should review every changed line and verify the boundary-hook output.

AI assistance: OpenAI Codex was used for local investigation, validation, and patch drafting.

### CODE REVIEW

**Comment by wuxibin89** [comment]
Reward should see both observation tokens and generated tokens.

**Comment by haoyang9804** [comment]
> Reward should see both observation tokens and generated tokens.

I think there is a gap. I agree that reward may need access to tool observations as context. What I’m trying to avoid here is treating tool observations as the model’s answer. In this patch solution_str is only the model-generated part, while the full decoded response is still available via extra_info["full_response_str"].
